# Big Data Project
## Expected outcome

In this notebook we will simulate expected outcome indicators

In [1]:
import os
import os.path
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from colorama import Fore

In [2]:
pathori = "../Data/Raw/Main/"
pathtemp = "../Data/Interim/"
pathdest = "../Data/Processed/"
filetemp = "main_competitions.csv"
filedest = "main_competitions.csv"

In [3]:
# Open dataset
url = pathtemp + filetemp
parse_dates=["Date"]
ds = pd.read_csv(url, parse_dates=parse_dates, index_col=False)

In [4]:
ds = ds.dropna()
display(ds.sample(10))

Country    Competition     Season  Div       Date        HomeTeam  \
62922        Italy         SerieA  2011-2012   I1 2011-10-16          Chievo   
20799      England        League1  2009-2010   E2 2009-12-19         Swindon   
13825      England     Conference  2009-2010   EC 2009-08-29  Kettering Town   
97962        Spain        Segunda  2011-2012  SP2 2011-11-19         Cordoba   
54232      Germany          Liga2  2009-2010   D2 2009-08-11        Duisburg   
72104  Netherlands     Eredivisie  2005-2006   N1 2005-05-11         Roda JC   
38375       France      Division2  2004-2005   F2 2005-01-04           Brest   
45001       France  LeChampionnat  2008-2009   F1 2009-05-16      St Etienne   
54995      Germany          Liga2  2011-2012   D2 2012-07-04        Duisburg   
13630      England     Conference  2008-2009   EC 2009-03-14  Kettering Town   

                 AwayTeam FTR   WHH   WHD   WHA   SBH   SBD    SBA   IWH  IWD  \
62922            Juventus   D  3.90  3.25  2.00  3.60  3.25  2.050  3.20  3.2   
20799            Brighton   H  1.83  3.20  3.75  2.00  3.25  3.400  1.90  3.2   
13825               Luton   D  2.50  3.30  2.38  2.70  3.40  2.250  2.60  3.1   
97962            Alcorcon   H  2.30  3.10  3.00  2.30  3.15  3.100  2.20  3.1   
54232               Ahlen   D  1.40  4.20  7.00  1.35  4.20  8.000  1.35  4.7   
72104           NAC Breda   D  1.95  3.50  3.10  2.00  3.25  3.400  1.85  3.2   
38375               Sedan   H  1.90  2.70  4.30  2.00  2.80  3.750  2.00  2.7   
45001            Toulouse   D  2.25  3.20  2.75  2.35  3.10  2.850  2.40  2.9   
54995       Ein Frankfurt   H  3.80  3.60  1.73  4.30  3.40  1.833  3.80  3.4   
13630  Eastbourne Borough   A  1.67  3.50  4.20  1.65  3.40  5.000  1.65  3.3   

        IWA   GBH   GBD   GBA  
62922  2.10  3.75  3.25  2.00  
20799  3.40  2.00  3.25  3.40  
13825  2.30  2.80  3.30  2.20  
97962  3.00  2.30  3.00  3.00  
54232  8.00  1.40  4.10  7.50  
72104  3.50  1.91  3.30  3.50  
38375  3.80  1.95  2.65  4.25  
45001  2.90  2.30  3.20  2.90  
54995  1.85  4.10  3.40  1.80  
13630  4.40  1.65  3.40  4.50

In [23]:
dssample = ds.loc[ds['Season']=='2011-2012'].loc[ds['Country']=='Spain'].loc[ds['AwayTeam']=='Barcelona'].loc[ds['HomeTeam']=='Granada']
dssample.head(50)

Country Competition     Season  Div       Date HomeTeam   AwayTeam FTR  \
92836   Spain     Primera  2011-2012  SP1 2011-10-25  Granada  Barcelona   A   

        WHH  WHD   WHA   SBH  SBD    SBA   IWH  IWD   IWA   GBH  GBD   GBA  
92836  23.0  7.0  1.14  19.0  8.5  1.111  14.0  6.2  1.15  17.0  8.0  1.12

**Success of a betting house**

In fact this indicator is the **'Accuracy'** because it is 'Success (TP and TN) related to total matches.

We expect to define a function like this: **accuracy (match, bethouse, scope)**

With params:
- match: Full information (register) of a match
- bethouse: Abrevation of the Bet House. Values: WH, SB, IW, GB
- scope: Columns to filter information. Values:  Country, Competition, Season, Team or combination of them.

This function will calculate accuracy = success/matches of a concret Bet House in the selected scope and in reference to the desired match.

Example: 
match = 

 	Country 	Competition 	Season 	Div 	Date 	HomeTeam 	AwayTeam 	FTR 	WHH 	WHD 	WHA 	SBH 	SBD 	SBA 	IWH 	IWD 	IWA 	GBH 	GBD 	GBA
    
0 	Belgium 	JupilerLeague 	2003-2004 	B1 	2003-08-08 	Club Brugge 	Genk 	H 	NaN 	NaN 	NaN 	1.44 	3.75 	6.5 	1.45 	3.8 	5.4 	1.4 	3.8 	6.85

result = accuracy(match, 'WH', ['Spain'])

result = 0.6


**Divergence**

In a individual match divergence it's a maesure of the difference between Bet House Quotes.

We expect to define a function like this: divergence(match)

We will use this function following these steps:

1. Filter matches of this week. Optional we can filter also by country or competition

2. Map rows calculating divergence

3. Reduce result amb obtain top N divergence values 

Result will be a Dataframe with the N rows with highest divergence, that they are the ones we need to pay attention to.


** Prediction Model **

Probably we will use a Classification Model, not selected yet, but that obtains a percentage of probability of win, draw or lose.

Example:
features = ['Country','Competition','Season','HomeTeam','AwayTeam','WHH','WHD','WHA','SBH','SBD','SBA','IWH','IWD','IWA','GBH','GBD','GBA']

label = 'FTR'

model = Model.trail(features, label)

prediction = model.predict(features)

Result:

Win: 'Belgium JupilerLeague 2003-2004 B1 2003-08-08 Club Brugge Genk  ...', 0.60

Drop: 'Belgium JupilerLeague 2003-2004 B1 2003-08-08 Club Brugge Genk  ...', 0.30
